sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/LocalShuffledRowRDD.scala

```diff
@@ -52,16 +51,48 @@ private final class LocalShuffledRowRDDPartition(
  */
 class LocalShuffledRowRDD(
      var dependency: ShuffleDependency[Int, InternalRow, InternalRow],
-     metrics: Map[String, SQLMetric])
+     metrics: Map[String, SQLMetric],
+     specifiedPartitionStartIndices: Option[Array[Int]] = None)
   extends RDD[InternalRow](dependency.rdd.context, Nil) {
 
   private[this] val numReducers = dependency.partitioner.numPartitions
   private[this] val numMappers = dependency.rdd.partitions.length
 
   override def getDependencies: Seq[Dependency[_]] = List(dependency)
 
-  override def getPartitions: Array[Partition] = {
+  /**
+   * To equally divide n elements into m buckets, basically each bucket should have n/m elements,
+   * for the remaining n%m elements, add one more element to the first n%m buckets each. Returns
+   * a sequence with length numBuckets and each value represents the start index of each bucket.
+   */
+  def equallyDivide(numElements: Int, numBuckets: Int): Seq[Int] = {
+    val elementsPerBucket = numElements / numBuckets
+    val remaining = numElements % numBuckets
+    val splitPoint = (elementsPerBucket + 1) * remaining
+    (0 until remaining).map(_ * (elementsPerBucket + 1)) ++
+      (remaining until numBuckets).map(i => splitPoint + (i - remaining) * elementsPerBucket)
+  }
+
+  private[this] val partitionStartIndices: Array[Int] = specifiedPartitionStartIndices match {
+    case Some(indices) => indices
+    case None =>
+      if (numReducers > numMappers) {
+        val numBuckets = numReducers / numMappers
+        equallyDivide(numReducers, numBuckets).toArray
+      } else {
+        Array(0)
+      }
+  }
+
+  private[this] val partitionEndIndices: Array[Int] =
+    Array.tabulate[Int](partitionStartIndices.length) { i =>
+      if (i < partitionStartIndices.length -1) {
+        partitionStartIndices(i + 1)
+      } else numReducers
+  }
 
+  override def getPartitions: Array[Partition] = {
+    assert(partitionStartIndices.length == partitionEndIndices.length)
     Array.tabulate[Partition](numMappers) { i =>
```

Comment:
This is problematic. The number of partitions should be "number of mappers" * "number of start/end index pairs"

Reply:
note that, one partition one task, if we don't update this place, we don't change the parallelism at all.